internal/controllers/machine/machine_controller.go

```diff
@@ -1025,3 +1088,193 @@ func (r *Reconciler) nodeToMachine(ctx context.Context, o client.Object) []recon
+	log := ctrl.LoggerFrom(ctx)
+
+	pvsWaitingForDetach := []*corev1.PersistentVolume{}
+	foundAttachedVolumeNames := sets.Set[string]{}
+	foundAttachedPVNames := sets.Set[string]{}
+
+	// List all PersistentVolumes and preserve the ones we have to wait for.
+	// Also store the found VolumeHandles and names of PersistentVolumes to check
+	// that all expected PersistentVolumes have been found.
+	// Note: pvsWaitingFor will not include PersistentVolumes we want to ignore.
+	// These are usually volumes which have a Pod which is expected to be kept.
+	persistentVolumeList := &corev1.PersistentVolumeList{}
+	for {
+		listOpts := []client.ListOption{
+			client.Continue(persistentVolumeList.GetContinue()),
+			client.Limit(100),
+		}
+		if err := c.List(ctx, persistentVolumeList, listOpts...); err != nil {
+			return nil, errors.Wrapf(err, "failed to list %T", persistentVolumeList)
+		}
+
+		for _, persistentVolume := range persistentVolumeList.Items {
+			found := false
+			// Lookup if the PersistentVolume matches an entry in attachedVolumeNames.
+			if persistentVolume.Spec.CSI != nil {
+				attachedVolumeName := fmt.Sprintf("kubernetes.io/csi/%s^%s", persistentVolume.Spec.CSI.Driver, persistentVolume.Spec.CSI.VolumeHandle)
+				if attachedVolumeNames.Has(attachedVolumeName) {
+					foundAttachedVolumeNames.Insert(attachedVolumeName)
+					found = true
+				}
+			}
+
+			// Lookup if the PersistentVolume matches an entry in attachedPVNames.
+			if attachedPVNames.Has(persistentVolume.Name) {
+				foundAttachedPVNames.Insert(persistentVolume.Name)
+				found = true
+			}
+
+			// PersistentVolume which do not match an entry in attachedVolumeNames or
+			// attachedPVNames can be ignored.
+			if !found {
+				continue
+			}
+
+			// Ignore PersistentVolumes which refer a PersistentVolumeClaim from pvcsToIgnore.
+			if persistentVolume.Spec.ClaimRef != nil && persistentVolume.Spec.ClaimRef.Kind == "PersistentVolumeClaim" {
+				pvcKey := types.NamespacedName{Namespace: persistentVolume.Spec.ClaimRef.Namespace, Name: persistentVolume.Spec.ClaimRef.Name}.String()
+				if pvcsToIgnore.Has(pvcKey) {
+					log.Info("Ignoring still attached PersistentVolume from undrained Pod", "PersistentVolume", klog.KObj(&persistentVolume), "PersistentVolumeClaim", klog.KRef(persistentVolume.Spec.ClaimRef.Namespace, persistentVolume.Spec.ClaimRef.Name))
+					continue
+				}
+			}
+
+			// Add the PersistentVolume to the list we are waiting for being detached.
+			pvsWaitingForDetach = append(pvsWaitingForDetach, &persistentVolume)
+		}
+
+		if persistentVolumeList.GetContinue() == "" {
+			break
+		}
+	}
+
+	errs := []error{}
+
+	// Check if for every entry in node.Status.AttachedVolumes, a corresponding PV was found, return an error otherwise.
+	attachedVolumeHandlesWithoutPV := attachedVolumeNames.Difference(foundAttachedVolumeNames)
+	if len(attachedVolumeHandlesWithoutPV) > 0 {
```

Comment:
How sure are we that every single entry in Node.status.attachedVolumes has to map to a PersistentVolume?

Because if we miss something, this will now always fail